src/treelearner/feature_histogram.hpp

```diff
@@ -850,11 +851,12 @@ class FeatureHistogram {
     }
   }
 
+  // FIXME Make constraints const again and have the cumulative constraints outside of it?
   template <bool USE_RAND, bool USE_MC, bool USE_L1, bool USE_MAX_OUTPUT, bool USE_SMOOTHING,
             bool REVERSE, bool SKIP_DEFAULT_BIN, bool NA_AS_MISSING>
   void FindBestThresholdSequentially(double sum_gradient, double sum_hessian,
                                      data_size_t num_data,
-                                     const ConstraintEntry& constraints,
+                                     FeatureConstraint* constraints,
```

Comment:
don't we really need to change this to an editable object?
I quick through the code, it seems the `CumulativeFeatureConstraint` is reset in each split. 
A simple solution is to make CumulativeFeatureConstraint `mutable`.

Reply:
Very good advice!